File: windows/src/Slipstream.Core/Identity/Fingerprint.cs
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;

namespace Slipstream.Core.Identity;

/// <summary>Spec §4: SHA-256 of the DER-encoded certificate, lowercase hex.</summary>
public static class Fingerprint
{
    public static string Of(X509Certificate2 certificate) => Of(certificate.RawData);

    public static string Of(ReadOnlySpan<byte> der) =>
        Convert.ToHexString(SHA256.HashData(der)).ToLowerInvariant();
}

File: android/core/src/main/kotlin/com/slipstream/core/identity/Fingerprint.kt
package com.slipstream.core.identity

import java.security.cert.X509Certificate
import java.security.MessageDigest

/**
 * Fingerprint is a lowercase hex SHA-256 of a certificate's DER encoding.
 * Always 64 characters (256 bits = 32 bytes = 64 hex digits).
 */
object Fingerprint {
    fun of(cert: X509Certificate): String {
        val derEncoded = cert.encoded
        val digest = MessageDigest.getInstance("SHA-256").digest(derEncoded)
        return digest.joinToString("") { "%02x".format(it) }
    }
}
